File: demo-swing/src/main/java/io/github/bric3/diaphanous/demo/DemoApp.java
/*
 * Diaphanous Swing
 *
 * Copyright (c) ${year} - ${name}
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package io.github.bric3.diaphanous.demo;

import io.github.bric3.diaphanous.MacToolbarStyle;
import io.github.bric3.diaphanous.MacWindowStyle;
import io.github.bric3.diaphanous.MacWindowStyler;

import javax.swing.JButton;
import javax.swing.JFrame;
import javax.swing.JLabel;
import javax.swing.JPanel;
import javax.swing.SwingUtilities;
import java.awt.BorderLayout;
import java.awt.Font;
import java.awt.GridBagLayout;
import java.util.concurrent.atomic.AtomicBoolean;

/**
 * Manual demo for toggling native macOS style attributes on a Swing frame.
 */
public final class DemoApp {
    private static final String APPLY_BUTTON_TEXT = "Apply macOS style";
    private static final String REVERT_BUTTON_TEXT = "Revert macOS style";
    private static final MacWindowStyle STYLED_WINDOW_STYLE = MacWindowStyle.builder()
        .transparentTitleBar(true)
        .fullSizeContentView(true)
        .titleVisible(false)
        .toolbarStyle(MacToolbarStyle.UNIFIED_COMPACT)
        .build();
    private static final MacWindowStyle STANDARD_WINDOW_STYLE = MacWindowStyle.builder()
        .transparentTitleBar(false)
        .fullSizeContentView(false)
        .titleVisible(true)
        .toolbarStyle(MacToolbarStyle.AUTOMATIC)
        .build();

    private DemoApp() {
    }

    /**
     * Starts the demo UI.
     *
     * @param args ignored
     */
    public static void main(String[] args) {
        SwingUtilities.invokeLater(DemoApp::showWindow);
    }

    private static void showWindow() {
        JFrame frame = new JFrame("Diaphanous Swing Demo");
        frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        frame.setSize(980, 640);
        frame.setLocationRelativeTo(null);

        JLabel title = new JLabel("Native macOS title bar restyling for this Swing window", JLabel.CENTER);
        title.setFont(new Font("SF Pro Text", Font.BOLD, 22));

        JLabel details = new JLabel("""
            <html>
            <div style='font-size: 15px; text-align: left;'>
              Clicking <b>Apply macOS style</b> will:
              <ul>
                <li>make the title bar transparent</li>
                <li>extend content into the title bar area (full-size content view)</li>
                <li>hide the text title</li>
                <li>switch to unified compact toolbar style</li>
                <li>click again to revert to a standard title bar</li>
              </ul>
            </div>
            </html>
            """);

        JButton applyStyleButton = new JButton(APPLY_BUTTON_TEXT);
        AtomicBoolean styledApplied = new AtomicBoolean(false);

        applyStyleButton.addActionListener(e -> {
            MacWindowStyle style = styledApplied.get() ? STANDARD_WINDOW_STYLE : STYLED_WINDOW_STYLE;
            MacWindowStyler.apply(frame, style);
            boolean nowStyled = !styledApplied.get();
            styledApplied.set(nowStyled);
            applyStyleButton.setText(nowStyled ? REVERT_BUTTON_TEXT : APPLY_BUTTON_TEXT);
        });

        JPanel centerPanel = new JPanel(new BorderLayout(0, 12));
        centerPanel.add(title, BorderLayout.NORTH);
        centerPanel.add(details, BorderLayout.CENTER);

        JPanel panel = new JPanel(new BorderLayout(16, 16));
        JPanel centered = new JPanel(new GridBagLayout());
        centered.add(centerPanel);
        panel.add(centered, BorderLayout.CENTER);
        panel.add(applyStyleButton, BorderLayout.SOUTH);

        frame.setContentPane(panel);
        frame.setVisible(true);
    }
}

File: demo-swing/src/main/kotlin/io/github/bric3/diaphanous/demo/DemoApp.kt
/*
 * Diaphanous Swing
 *
 * Copyright (c) 2026 - Brice Dutheil
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package io.github.bric3.diaphanous.demo

import io.github.bric3.diaphanous.MacosStartupReveal
import io.github.bric3.diaphanous.backdrop.RootErasingContentPane
import io.github.bric3.diaphanous.backdrop.WindowBackdrop
import io.github.bric3.diaphanous.decorations.MacosWindowAppearanceSpec
import io.github.bric3.diaphanous.decorations.WindowPresentations
import java.awt.BorderLayout
import java.awt.Color
import java.awt.Dimension
import java.awt.Font
import java.awt.GridBagLayout
import java.awt.LayoutManager
import javax.swing.JComponent
import javax.swing.JFrame
import javax.swing.JLabel
import javax.swing.JPanel
import javax.swing.SwingUtilities

/**
 * Manual demo for toggling native macOS style attributes on a Swing frame.
 */
object DemoApp {
    /**
     * Starts the demo UI.
     */
    @JvmStatic
    fun main(args: Array<String>) {
        configureApplicationAppearancePolicy()
        val options = parseOptions(args)
        SwingUtilities.invokeLater { showWindow(options) }
    }

    private fun configureApplicationAppearancePolicy() {
        val isMac = System.getProperty("os.name", "").contains("Mac", ignoreCase = true)
        if (!isMac) {
            return
        }
        val key = "apple.awt.application.appearance"
        if (System.getProperty(key).isNullOrBlank()) {
            System.setProperty(key, "system")
        }
    }

    private fun parseOptions(args: Array<String>): LaunchOptions {
        var appearance = MacosWindowAppearanceSpec.SYSTEM

        for (arg in args) {
            when {
                arg.startsWith("--appearance=", ignoreCase = true) -> {
                    val raw = arg.substringAfter('=')
                    appearance = parseAppearance(raw) ?: appearance
                }
            }
        }
        return LaunchOptions(appearance)
    }

    private fun parseAppearance(raw: String): MacosWindowAppearanceSpec? {
        val normalized = raw.trim().replace('-', '_').uppercase()
        return MacosWindowAppearanceSpec.entries.firstOrNull { it.name == normalized }
    }

    private fun showWindow(options: LaunchOptions) {
        val frame = JFrame("Diaphanous Swing Demo").apply {
            defaultCloseOperation = JFrame.EXIT_ON_CLOSE
            setSize(980, 640)
            setLocationRelativeTo(null)
            rootPane.isOpaque = false
            layeredPane.isOpaque = false
        }

        val title = JLabel("Native macOS experimental backdrop for this Swing window", JLabel.CENTER).apply {
            font = Font("SF Pro Text", Font.BOLD, 22)
        }


        val windowBackdropControls = WindowBackdropControls(frame) {
            WindowBackdrop.apply(frame, it, WindowPresentations.isCompatibleWithBackdropPredicate())
        }
        val windowDecorationsControls = WindowDecorationsControls(
            initialAppearance = options.appearance,
            onStyleChange = { style -> WindowPresentations.applyDecorations(frame, style) },
            onAppearanceChange = { appearance ->
                WindowPresentations.applyAppearance(frame, appearance)
                WindowBackdrop.apply(frame, windowBackdropControls.currentSpec(),
                    WindowPresentations.isCompatibleWithBackdropPredicate())
            }
        )
        val topSeriesPanel = RandomTimeseriesPanel()
        val topTimeseriesColorControls = TopTimeseriesColorControls(topSeriesPanel)

        val settingsPanel = transparentPanel(BorderLayout(0, 8)).apply {
            add(windowDecorationsControls.component, BorderLayout.NORTH)
            add(windowBackdropControls.component, BorderLayout.CENTER)
            preferredSize = Dimension(320, preferredSize.height)
        }

        val centerPanel = transparentPanel(BorderLayout(0, 0)).apply {
            add(title, BorderLayout.NORTH)
            add(
                transparentPanel(BorderLayout(0, 16)).apply {
                    add(topTimeseriesColorControls.component, BorderLayout.WEST)
                    add(settingsPanel, BorderLayout.EAST)
                },
                BorderLayout.CENTER
            )
        }

        frame.contentPane = RootErasingContentPane(BorderLayout(16, 16)).apply {
            add(topSeriesPanel, BorderLayout.NORTH)
            add(
                transparentPanel(GridBagLayout()).apply {
                    add(centerPanel)
                },
                BorderLayout.CENTER
            )
        }
        WindowPresentations.applyAppearance(frame, windowDecorationsControls.currentAppearanceSpec())
        WindowPresentations.applyDecorations(frame, windowDecorationsControls.currentDecorationsSpec())
        WindowBackdrop.apply(frame, windowBackdropControls.currentSpec(),
            WindowPresentations.isCompatibleWithBackdropPredicate())
        MacosStartupReveal.show(frame)
        if (java.lang.Boolean.getBoolean("diaphanous.dump.swing")) {
            dumpComponentTree(frame.rootPane, 0)
        }
    }

    fun transparentPanel(layout: LayoutManager): JPanel = JPanel(layout).apply {
        isOpaque = false
        background = Color(0, 0, 0, 0)
    }

    private data class LaunchOptions(
        val appearance: MacosWindowAppearanceSpec
    )

    private fun dumpComponentTree(component: java.awt.Component, depth: Int) {
        val indent = "  ".repeat(depth)
        val opaque = if (component is JComponent) component.isOpaque else false
        val background = component.background
        val namePart = if (component.name != null) " name=${component.name}" else ""
        println(
            "$indent- ${component.javaClass.simpleName}$namePart visible=${component.isVisible} " +
                    "opaque=$opaque bg=rgba(${background.red},${background.green},${background.blue},${background.alpha})"
        )
        if (component is java.awt.Container) {
            component.components.forEach { child -> dumpComponentTree(child, depth + 1) }
        }
    }
}
